File: src/main/java/ru/ioleksiv/telegram/bot/core/controller/updater/Updater.java
package ru.ioleksiv.telegram.bot.core.controller.updater;

import org.springframework.stereotype.Controller;
import ru.ioleksiv.telegram.bot.api.controller.TelegramProcessor;
import ru.ioleksiv.telegram.bot.api.controller.TelegramUpdater;
import ru.ioleksiv.telegram.bot.api.model.objects.Update;
import ru.ioleksiv.telegram.bot.core.controller.network.Loader;

@Controller
public class Updater implements TelegramUpdater {
    private final TelegramProcessor telegramProcessor;
    private final Loader loader;

    public Updater(TelegramProcessor telegramProcessor, Loader loader) {
        this.telegramProcessor = telegramProcessor;
        this.loader = loader;
    }

    @Override
    public final void longPolling() {

        Iterable<Update> updates = loader.loadUpdates();

        for (Update update : updates) {
            try {
                telegramProcessor.receive(update);
            }
            catch (Exception processException) {
                error(update, processException);
            }
        }

    }

    @Override
    public void error(Update update, Exception onProcessException) {
        // stub
    }
}

File: core/src/main/kotlin/io/github/oleksivio/tl/kbot/core/core/controller/updater/Updater.kt
package io.github.oleksivio.tl.kbot.core.core.controller.updater

import io.github.oleksivio.tl.kbot.core.controller.TelegramProcessor
import io.github.oleksivio.tl.kbot.core.controller.TelegramUpdater
import io.github.oleksivio.tl.kbot.core.controller.UpdateErrorListener
import io.github.oleksivio.tl.kbot.core.core.controller.network.Loader

class Updater(
    private val telegramProcessor: TelegramProcessor,
    private val loader: Loader
) : TelegramUpdater {

    private var updateErrorListener: UpdateErrorListener = { _, e -> e.printStackTrace() }

    override fun longPolling() {

        val updates = loader.loadUpdates()

        for (update in updates) {
            try {
                telegramProcessor.receive(update)
            } catch (processException: Exception) {
                updateErrorListener(update, processException)
            }
        }
    }

    fun setErrorListener(listener: UpdateErrorListener) {
        this.updateErrorListener = listener
    }
}
